dialectic-compiler/src/lib.rs

```diff
@@ -0,0 +1,909 @@
+use {
+    lazy_static::lazy_static,
+    proc_macro2::{Span, TokenStream},
+    quote::{quote, ToTokens},
+    std::{fmt, ops, rc::Rc},
+    syn::{spanned::Spanned, Error, Ident, Type},
+    thiserror::Error,
+    thunderdome::{Arena, Index},
+};
+
+pub mod parse;
+
+#[derive(Error, Debug, Clone)]
+pub enum CompileError {
+    #[error("label name `'{0}` shadows a label name that is already in scope")]
+    ShadowedLabel(String),
+    #[error("undeclared label `'{0}`")]
+    UndeclaredLabel(String),
+    #[error("cannot `continue` outside of a loop")]
+    ContinueOutsideLoop,
+    #[error("cannot `break` outside of a loop")]
+    BreakOutsideLoop,
+}
+
+#[derive(Debug, Clone)]
+pub struct SpannedCompileError {
+    span: Span,
+    kind: CompileError,
+}
+
+/// A shim for parsing the root level of a macro invocation, so
+/// that a session type may be written as a block w/o an extra
+/// layer of braces.
+#[derive(Debug, Clone)]
+pub struct Invocation {
+    pub syntax: SyntaxNode,
+}
+
+#[derive(Debug, Clone)]
+pub struct SyntaxNode {
+    pub expr: Syntax,
+    pub span: Span,
+}
+
+#[derive(Debug, Clone)]
+pub enum Syntax {
+    Recv(Type),
+    Send(Type),
+    Call(Box<SyntaxNode>),
+    Choose(Vec<SyntaxNode>),
+    Offer(Vec<SyntaxNode>),
+    Split(Box<SyntaxNode>, Box<SyntaxNode>),
+    Loop(Option<String>, Box<SyntaxNode>),
+    Break(Option<String>),
+    Continue(Option<String>),
+    Block(Vec<SyntaxNode>),
+    Type(Type),
+}
+
+#[derive(Debug, Clone)]
+struct CfgNode {
+    expr: Ir,
+    next: Option<Index>,
+    span: Span,
+}
+
+#[derive(Debug, Clone)]
+enum Ir {
+    Done,
+    Recv(Type),
+    Send(Type),
+    Call(Index),
+    Choose(Vec<Index>),
+    Offer(Vec<Index>),
+    Split(Index, Index),
+    Loop(Option<String>, Index),
+    IndexedBreak(usize),
+    IndexedContinue(usize),
+    LabeledBreak(String),
+    LabeledContinue(String),
+    Type(Type),
+    Error,
+}
+
+#[derive(Clone, Debug)]
+pub enum Target {
+    Done,
+    Recv(Type, Rc<Target>),
+    Send(Type, Rc<Target>),
+    Choose(Vec<Target>),
+    Offer(Vec<Target>),
+    Loop(Rc<Target>),
+    Continue(usize),
+    Split(Rc<Target>, Rc<Target>),
+    Call(Rc<Target>, Rc<Target>),
+    Then(Rc<Target>, Rc<Target>),
+    Type(Type),
+    Error,
```

Comment:
See comment about line 82.

Reply:
This one actually does have a bit of a reason... if there's, say, an error where a loop body would get turned into an error node because of a bad label/shadowing label, then turning that loop node into an error node would lose all error information about errors inside the loop body. So if we want to emit all errors we have to do it w/ a scheme where at least some errors exist outside the AST. I think that's the same reason I used an empty error node for the `Ir` enum too.